src/js/diff.js

```diff
@@ -0,0 +1,549 @@
+ *
+ */
+gpii.diff.compareArrays = function (leftArray, rightArray) {
+    if (!Array.isArray(leftArray) || !Array.isArray(rightArray)) {
+        return gpii.diff.singleValueDiff(leftArray, rightArray);
+    }
+    else {
+        var segments = [];
+
+        if (gpii.diff.arraysEqual(leftArray, rightArray)) {
+            segments.push({ arrayValue: leftArray, type: "unchanged"});
+        }
+        else {
+            // Find the longest match, then process the remaining "tail" and "head", as with the string engine above.
+            var longestCommonSegmentDef = gpii.diff.longestCommonArraySegment(leftArray, rightArray);
+            if (longestCommonSegmentDef.segment.length) {
+                // Compare the "leaders"
+                var leftLeader = leftArray.slice(0, longestCommonSegmentDef.leftIndex);
+                var rightLeader = rightArray.slice(0, longestCommonSegmentDef.rightIndex);
+                if (leftLeader.length || rightLeader.length) {
+                    var leadingSegments = gpii.diff.compareArrays(leftLeader, rightLeader);
+                    segments = segments.concat(leadingSegments);
+                }
+
+                // Add the longest segment itself
+                segments.push({ arrayValue: longestCommonSegmentDef.segment, type: "unchanged"});
+
+                // Compare the "trailers"
+                var leftTrailer = leftArray.slice(longestCommonSegmentDef.leftIndex + longestCommonSegmentDef.segment.length);
+                var rightTrailer = rightArray.slice(longestCommonSegmentDef.rightIndex + longestCommonSegmentDef.segment.length);
+                if (leftTrailer.length || rightTrailer.length) {
+                    var trailingSegments = gpii.diff.compareArrays(leftTrailer, rightTrailer);
+                    segments = segments.concat(trailingSegments);
+                }
+            }
+            else {
+                return gpii.diff.singleValueDiff(leftArray, rightArray, true);
+            }
+        }
+
+        return segments;
+    }
+};
+
+/***
+ *
+ * A function that returns the (first) longest matching segment found in two arrays.  We use this to tree out from the
+ * best match within an array, so that we produce a change report with the least changes possible.
+ *
+ * @param leftArray - An array of values.
+ * @param rightArray - An array of values to compare to `leftArray`.
+ * @return {Object} - See below.
+ *
+ * The object returned has three keys:
+ *
+ * `leftIndex`:  The position of the matching segment in `leftArray`.  If there is no matching segment, this will be `-1`.
+ * `rightIndex`: The position of the matching segment in `rightArray`.  If there is no matching segment, this will be `-1`.
+ * `segment`:    The longest set of matching elements (in order) common to both arrays.  If there are multiple sets of the same length, the first is returned.
+ *
+ */
+gpii.diff.longestCommonArraySegment = function (leftArray, rightArray) {
```

Comment:
The comment should note some implementation subtleties, which should perhaps even be surfaced to top-level comments. Whilst exact equality throughout this algorithm is done on a deep equality basis, this algorithm uses only shallow equality as operated by "indexOf" as invoked on line 378. This means that situations like
leftArray: [{a: 1}, {b: 2}, 3]
rightArray: [{a: 1}, {b: 2}]
will be treated as a total diff. I don't propose fixing this right now but the user should be clear about the quality of support. Also it should be noted that this is a quadratic-time algorithm which may not be appropriate for very long arrays. Algorithms do exist which can compute this segment in O(n log n) time but again I am not proposing that we adopt one - but all the same the quality of implementation should be clearly flagged to the user.

Reply:
After a little research: It should be noted that this is actually a *cubic* time algorithm (taking into account the costs of the arraysEqual call on line 381) - it could be reduced to quadratic time by implementing the standard LCS algorithm described on Wikipedia, or to n log n by some heroic efforts involving the "Method of Four Russians" ...

Reply:
I have added that as a test case and will work on an approach that handles it better. 

As these "quadratic time" functions are called implicitly by various other functions, I will have to add a suitable warning throughout.

Reply:
The new LCS approach specifically supports the case mentioned above.